### PR TITLE
[dynamo] New logging artifacts for source code attribution

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104676
* #104672
* __->__ #104013

Prototype for the feature request:
>When working on a codebase that is unfamiliar to you, it can be helpful to single step through all of the code to see what is getting executed, what conditional branches are taken, and where indirect function jumps go. Model x-ray uses dynamo to give you a single step log of every source code line that does something relevant (i.e., a Tensor operation)

Dynamo logs to the ~`starts_line`~ `trace_source` logging artifact at the start of tracing new bytecode with a new line. It logs the line of source code associated with that bytecode.

~~Dynamo logs to the `graph_source` logging when a FX GraphModule is constructed. For each node in the graph, it logs the location of the original source code associated with that node.~~

Development notes: https://docs.google.com/document/d/1LjFeHzCgDDt535QUq5HydcQs56d7jWl5RvW8TLZN19g/edit?usp=sharing

Since the draft, we removed the `graph_source` logging artifact since printing the code of `GraphModule`s already displays the original source.

Sample:

```python
import torch
from functorch.experimental.control_flow import cond

def true_fn(x):
    return x * 2

def false_fn(x):
    return x * 3

def f_cond(pred, x):
    return cond(pred, true_fn, false_fn, [x])

def f_outer(pred, x):
    y = f_cond(pred, x)
    if x.sum() > 0:
        x = x * 2
    else:
        x = x * 3
    return x, y

opt_f_cond = torch.compile(f_outer, backend="eager")
opt_f_cond(torch.tensor(True), torch.randn(3, 3))
```

Logs:
```shell
$ TORCH_LOGS="trace_source" python playground8.py 
TRACE starts_line f_outer playground8.py:54
    def f_outer(pred, x):
TRACE starts_line f_outer playground8.py:55
        y = f_cond(pred, x)
TRACE starts_line f_cond playground8.py:51 (inline depth: 1)
    def f_cond(pred, x):
TRACE starts_line f_cond playground8.py:52 (inline depth: 1)
        return cond(pred, true_fn, false_fn, [x])
TRACE starts_line true_fn playground8.py:45 (inline depth: 2)
    def true_fn(x):
TRACE starts_line true_fn playground8.py:46 (inline depth: 2)
        return x * 2
TRACE starts_line false_fn playground8.py:48 (inline depth: 2)
    def false_fn(x):
TRACE starts_line false_fn playground8.py:49 (inline depth: 2)
        return x * 3
TRACE starts_line f_outer playground8.py:56
        if x.sum() > 0:
TRACE starts_line <resume in f_outer> playground8.py:56
        if x.sum() > 0:
TRACE starts_line <resume in f_outer> playground8.py:57
            x = x * 2
TRACE starts_line <resume in f_outer> playground8.py:60
        return x, y
```

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78

### CODE REVIEW

**Comment by ezyang** [comment]
Nice! Some things I observe when looking at the logs

* We sometimes skip backwards a line number. This is consistent with pdb, since when we step, e.g., an if statement, we first process the overall if statement, then we do the inner expressions, then we evaluate the entire expression. But it's kind of hard to explain to naive users. When we doc this, probably explaining how this relates to pdb probably will be the best way to explain it. It would be nice if we could get column spans but we only really can support them in sufficiently recent CPython
* The log lines are soooo long. This is more of an infra problem but I wonder if there's a way we can suppress `[2023-06-22 01:15:16,414] torch._dynamo.symbolic_convert.__starts_line: [DEBUG]` for these messages, it really is unnecessary
* I'm ambivalent about two lines per log entry. On the one hand, it ensures the indentations for source prints all line up, which is good. On the other hand, it would have been nice to only have one line per log. The only alternative is something like `source code  # file:lineno` but IDK if that actually looks better. Up to you.

File: torch/_logging/_registrations.py [hunk 1]
```diff
@@ -11,6 +11,8 @@
 register_artifact("bytecode", off_by_default=True)
 register_artifact("graph")
 register_artifact("graph_code")
+register_artifact("graph_source", off_by_default=True)
+register_artifact("starts_line")
```
Comment (ezyang) [inline torch/_logging/_registrations.py:15]: I'm not sure this is a very explicable name for end users. How about something like "source_trace"?

File: torch/_dynamo/symbolic_convert.py [hunk 2]
```diff
@@ -599,6 +602,23 @@ def inline_user_function_return(self, fn, args, kwargs):
...
+            "TRACE starts_line %s%s %s:%s",
+            inline_depth_str,
+            self.f_code.co_name,
+            self.f_code.co_filename,
+            self.lineno,
+        )
+        line = linecache.getline(self.f_code.co_filename, self.lineno)
+        if line:
+            # strip the newline at the end
+            line = line[:-1]
```
Comment (ezyang) [inline torch/_dynamo/symbolic_convert.py:619]: If the file is missing trailing newline, line isn't guaranteed to have a newline, right? Maybe rstrip() is better.

File: torch/_dynamo/symbolic_convert.py [hunk 3]
```diff
@@ -599,6 +602,23 @@ def inline_user_function_return(self, fn, args, kwargs):
...
+        starts_line_log.debug("    %s", (" " * 4 * self.inline_depth) + line)
```
Comment (ezyang) [inline torch/_dynamo/symbolic_convert.py:620]: As written, this always does the string formatting even if the debug log is ignored. The formatting here is pretty complicated, maybe best to test log level and short circuit out.

File: torch/_dynamo/__init__.py [hunk 4]
```diff
@@ -64,6 +64,10 @@ def reset() -> None:
     compilation_metrics.clear()
     reset_frame_count()
 
+    import linecache
+
+    linecache.clearcache()
```
Comment (ezyang) [inline torch/_dynamo/__init__.py:69]: hmm?

File: torch/_dynamo/symbolic_convert.py [hunk 5]
```diff
@@ -2278,6 +2304,7 @@ def __init__(
             code_options={k: getattr(code, k) for k in dir(code)},
             f_code=code,
             export=parent.export,
+            inline_depth=parent.inline_depth + 1,
```
Comment (voznesenskym) [inline torch/_dynamo/symbolic_convert.py:2307]: Can you make sure this works correctly w/ the higher order ops nested output_graph codepaths?

**Comment by voznesenskym** [comment]
1) What does it look like if I have a really long line with a lot of calls on it? ex `f = a(b(b(c.d, c.e, c.f))) + g(h, i)` your example code is super terse and short so its hard to extrapolate it to real code

2) Why do we need this? we already have logs like `[2023-06-22 00:50:57,853] torch._dynamo.symbolic_convert: [DEBUG] TRACE starts_line /scratch/voz/work/pytorch/torch/distributed/fsdp/_runtime_utils.py:229` - isn't this just noise? I can always go look at the code right?

File: torch/_dynamo/__init__.py (hunk 4 above)
Comment (williamwen42) [reply torch/_dynamo/__init__.py:69]: It makes sense to me that we can clear the linecache when we reset dynamo. Are you suggesting that we do not?

**Comment by williamwen42** [comment]
@voznesenskym 
1. It would unfortunately print the entire line, but this seems to be a limitation related to the language. In Python < 3.11, we simply do not have information in bytecode instructions to pinpoint exactly which part of the line corresponds to the bytecode. In 3.11+, there are now more fields available to us (i.e. `end_lineno`, `col_offset`, `end_col_offset`) so that we can determine which sections of lines correspond to our bytecode. We could add this to our logging as a 3.11+ feature.
2. The point is to put all the existing `TRACE starts_line` logs in one logging artifact and also output the source code, so that they're easier to look at. And the purpose of the `graph_source` logging artifact is to see which lines of code are relevant in the resulting FX graph.

**Comment by ezyang** [comment]
Changing the formatter would be my preference!

File: torch/_dynamo/__init__.py (hunk 4 above)
Comment (ezyang) [reply torch/_dynamo/__init__.py:69]: I mean... it doesn't seem harmful but OTOH is this actually going to do anything?

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**